jaxrs-client/src/main/java/io/micronaut/jaxrs/client/JaxRsInvocation.java

```diff
@@ -0,0 +1,525 @@
+        }
+    }
+
+    <T> T invokeExchange(Argument<T> type, Entity<?> entity) {
+        return asyncBlock(async(method, type, entity));
+    }
+
+    private <T> CompletableFuture<T> async(Argument<T> type) {
+        return async((String) null, type, null);
+    }
+
+    private <T> CompletableFuture<T> async(HttpMethod method, Argument<T> type) {
+        return async(method.name(), type, null);
+    }
+
+    private <T> CompletableFuture<T> async(String method, Argument<T> type) {
+        return async(method, type, null);
+    }
+
+    private <T> CompletableFuture<T> async(HttpMethod method, Argument<T> type, Entity<?> entity) {
+        return async(method.name(), type, entity);
+    }
+
+    private <T> CompletableFuture<T> async(String method, Argument<T> type, Entity<?> entity) {
+        CompletableFuture<T> future = new CompletableFuture<>();
+        try {
+            Argument<Object> requestBodyType = Argument.of(Object.class);
+            if (entity == null) {
+                entity = this.entity;
+            }
+            MutableHttpRequest<Object> request = createRequest(method, entity);
+            if (entity != null) {
+                Object entityValue = entity.getEntity();
+                if (entityValue instanceof GenericEntity<?> genericEntity) {
+                    requestBodyType = (Argument<Object>) ArgumentUtil.from(genericEntity);
+                } else {
+                    requestBodyType = (Argument<Object>) Argument.of(entityValue.getClass(), ArgumentUtil.createAnnotationMetadata(entity.getAnnotations()));
+                }
+            }
+            List<ClientRequestFilter> requestFilters = configuration.getRequestFilters();
+            JaxRsClientRequestContext requestContext = new JaxRsClientRequestContext(client, configuration, request, requestBodyType);
+            if (!requestFilters.isEmpty()) {
+                for (ClientRequestFilter requestFilter : requestFilters) {
+                    requestFilter.filter(requestContext);
+                    Response response = requestContext.getResponse();
+                    if (response != null) {
+                        response = filterResponse(response, requestContext);
+                        if (type.getType().equals(Response.class)) {
+                            future.complete((T) response);
+                        } else {
+                            if (response.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL) {
+                                future.complete(response.readEntity(type.getType()));
+                            } else {
+                                future.completeExceptionally(new WebApplicationException(response));
+                            }
+                        }
+                        return future;
+                    }
+                }
+            }
+            client.getHttpClient().exchange(request)
+                .subscribe(new Subscriber<>() {
```

Comment:
maybe easier to implement using reactor instead of a custom Subscriber?

Reply:
I don't think it's a big deal. It would be nice to have async api for the client